### PR TITLE
Add many-to-many relationship between Society and User, add input validation to admin page

### PR DESCRIPTION
**Pull request by dnlburgess**
- Add an association table to the database, so societies can have multiple admins who can administer multiple societies.
- Add input validation to /<short_name>/admin page 
- Use SRCF styles on admin page
- Display a list of current admins on admin page and allow adding new or existing users as admins (note admins cannot be removed yet - do we want this?)

### CODE REVIEW

File: lightbluetent/admin.py [hunk 1]
```diff
@@ -1,39 +1,194 @@
...
+        # add this society to that user's list of societies
+
+        # TODO: identifying users should be CRSid-based, changed to email address for testing purposes
+        existing_user = User.query.filter_by(email=values["email_address"]).first()
+        if existing_user:
+            # Check the existing user isn't already an admin
+            if not society in existing_user.societies:
+                pass
+            else:
+                errors["crsid"] = "This user is already an administrator."
```
Comment (edwinbalani) [inline lightbluetent/admin.py:140]: I think this section could benefit from rework as follows:

  * The first 'if not x: pass; else: do_y().` block is a bit confusing.  `if x: do_y()` would make more sense.
  * I think if the user is already an administrator there's no point validating the other bits of input (or indeed displaying validation errors for them), so we can skip forward to rendering the template.  I'd usually be annoyed by a form that doesn't show validation errors across all of its input, but perhaps we can make an exception here?

Having said all that, perhaps what we really want is a two-page flow: the first just asks for the CRSid.  Then it pops open a generic 'edit' page, using existing user data if we have it or otherwise leaving the form blank.  Appropriate UI prompts can make it clear what's happening e.g. "This person is already an admin, here are their details.  Edit them as you wish then save your changes".

File: lightbluetent/admin.py [hunk 2]
```diff
@@ -1,39 +1,194 @@
...
+
+            # Check that the values we already have have been given by the registrant.
+            if existing_user.first_name != values["first_name"]:
+                errors["first_name"] = "This CRSid is already registered, but not under this name. Please use the user's registered name."
+            if existing_user.surname != values["surname"]:
+                errors["surname"] = "This CRSid is already registered, but not under this name. Please use the user's registered name."
+            if existing_user.email != values["email_address"]:
+                errors["email_address"] = "This CRSid is already registered, but not under this email address. Please use the user's registered email address."
```
Comment (edwinbalani) [inline lightbluetent/admin.py:148]: Is this a good idea?  I don't really see the point in making sure that someone's first name, surname, email address as they were previously entered into LBT (including typos and extra whitespace) match exactly on addition.

I asked in chat but will ask here too for coverage - what's the reason for checking these if the user exists?  It might be less surprising if we just ignore the details given here if the user already exists.

File: lightbluetent/admin.py [hunk 3]
```diff
@@ -1,39 +1,194 @@
-from flask import Blueprint, render_template, request, flash
-from lightbluetent.models import db
+import re
+
+from flask import Blueprint, render_template, request, flash, abort, redirect, url_for
+from lightbluetent.models import db, User, Society
+from lightbluetent.utils import auth_decorator, validate_email
+
+ILLEGAL_NAME_RE = re.compile(r'[:,=\n]')
+ILLEGAL_NAME_ERR = "Please do not use any of the following characters: : , = ⏎ "
```
Comment (edwinbalani) [inline lightbluetent/admin.py:8]: I can see why newlines should be disallowed, but why the rest?

File: lightbluetent/templates/admin/admin.html [hunk 4]
```diff
@@ -2,42 +2,123 @@
 
 {% block body %}
 
-  <p>Change your stand's settings:</p>
+  <p><b>Current administrators:</b></p>
+  {%- for admin in admins %}
+    <p><b>{{ admin.crsid }}</b> ({{ admin.first_name }} {{ admin.surname }} &lt;{{ admin.email }}&gt;)</p>
+  {%- endfor %}
+  <a href="{{ url_for("admin.add", short_name=short_name) }}">Add an administrator</a>
```
Comment (edwinbalani) [inline lightbluetent/templates/admin/admin.html:9]: Is there a mechanism to remove administrators?

File: lightbluetent/templates/admin/admin_add.html [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      <label class="col-lg-2 col-md-3 col-sm-4 col-form-label" for="crsid">CRSid</label>
+      <div class="col-lg-10 col-md-9 col-sm-8">
+        <input type="text" name=crsid id="crsid" class="form-control{% if errors.crsid is defined %} is-invalid{% endif %}" value="{{ crsid }}" required>
+        {%- if errors.crsid is defined %}
+          <small class="invalid-feedback">{{ errors.crsid }}</small>
+        {%- endif %}
+      </div>
+    </div>
+
+    <small>Please ensure the moderator you're adding has given permission for you to share their data with us!</small>
```
Comment (edwinbalani) [inline lightbluetent/templates/admin/admin_add.html:52]: Choice of language nit: for consistency this might be better as "Please ensure the administrator ...".

Or maybe we should reverse it and make everything else say 'moderator'?

**Review by edwinbalani — CHANGES_REQUESTED** [review]
This looks like it mostly does the job! :)

I was a bit confused about the input validation for the add-admin page; comments added.  I've also queried some of the input restrictions we have that aren't immediately clear to me.

Other than that, just minor things commented on.

File: lightbluetent/templates/home/register.html [hunk 6]
```diff
@@ -11,7 +11,7 @@
         {%- if errors.first_name is defined %}
                 <small class="invalid-feedback">{{ errors.first_name }}</small>
         {%- endif %}
-        <small class="form-text text-muted">Please only use unaccented English characters in your name.</small>
+        <small class="form-text text-muted">Please use only the unaccented characters Aa - Zz.</small>
```
[the remark this answers is not in the document]
Comment (dnlburgess) [reply lightbluetent/templates/home/register.html:14]: Yeah, not 100% happy about this either. No technical restriction as far as I know. I originally took this directly from control-panel (https://github.com/SRCF/control-panel/blob/master/control/templates/signup/signup.html#L34) but something about the reference to the 'English' alphabet didn't sit right with me. Surveyed a few friends and was told that asking for 'Latin' characters might confuse people, hence I went with this for now. Thoughts?

If it's fine to use a wider range of characters then I suppose we can do that!

File: lightbluetent/admin.py (hunk 3 above)
Comment (dnlburgess) [reply lightbluetent/admin.py:8]: Again, just taken from control-panel: will remove if everyone's happy to.

Comment (edwinbalani) [reply lightbluetent/admin.py:8]: Ah, I see -- the reason we have those in control-panel is so that the [GECOS field](https://en.wikipedia.org/wiki/Gecos_field) in /etc/passwd doesn't receive invalid data.  It can definitely go here.

File: lightbluetent/templates/home/register.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (edwinbalani) [reply lightbluetent/templates/home/register.html:14]: That will probably have been because a socname in control panel will become a Unix username, and if you have anything too exotic in there it'll upset a lot of systems.  (We don't allow numbers for socs to avoid 99.9999% of possible CRSid collisions, since all but the very oldest of CRSids contain numbers.)  It'd also feed into the `SOCNAME.soc.srcf.net` domain, which could indeed be punycoded to allow non-English characters, although we don't have the mechanisms for that as it stands.

In conclusion, I think we can and should be more liberal here, since this value is will just feed into the 'path' part of some URL (which can get URL-encoded as necessary).
